File: ai/target_spec.py
"""Validated TargetSpec contract shared with the on-device CV selector."""

from __future__ import annotations

import json
import math
from collections.abc import Mapping
from dataclasses import dataclass
from enum import StrEnum
from pathlib import Path
from typing import Any, TypeVar

from ai.taxonomy import OBJECTS365_YOLO26

EnumValue = TypeVar("EnumValue", bound=StrEnum)
SUPPORTED_SCHEMA_VERSIONS = frozenset({"0.1", "0.2"})


class TargetSpecStatus(StrEnum):
    OK = "ok"
    NEEDS_CLARIFICATION = "needs_clarification"
    FAILED = "failed"


class SubjectType(StrEnum):
    PERSON = "person"
    OBJECT = "object"
    LANDSCAPE = "landscape"
    # 서류·종이·신분증 (2026-08-30) — bbox 조준 대상이 없고 앱이 텍스트 영역으로 프레이밍한다.
    # 사진은 기기 밖으로 나가지 않는다(앱이 업로드를 생략).
    DOCUMENT = "document"


class Framing(StrEnum):
    CLOSEUP = "closeup"
    FULL_BODY = "full_body"
    WIDE = "wide"


class TargetSpecSource(StrEnum):
    CLOVA = "clova"
    ELEVENLABS = "elevenlabs"
    ONDEVICE = "ondevice"


@dataclass(frozen=True, slots=True)
class TargetSpec:
    """Machine-validated representation of ``ai/target_spec_schema.md`` v0.1/v0.2."""

    session_id: str
    raw_text: str
    source: TargetSpecSource
    schema_version: str = "0.1"
    status: TargetSpecStatus = TargetSpecStatus.OK
    subject_type: SubjectType = SubjectType.PERSON
    subject_count: int | None = None
    framing: Framing = Framing.FULL_BODY
    confidence: float = 0.0
    object_label: str | None = None

    def __post_init__(self) -> None:
        if self.schema_version not in SUPPORTED_SCHEMA_VERSIONS:
            raise ValueError(f"Unsupported TargetSpec schemaVersion: {self.schema_version}")
        if not isinstance(self.session_id, str):
            raise TypeError("TargetSpec sessionId must be a string")
        if not self.session_id.strip():
            raise ValueError("TargetSpec sessionId must be a non-empty string")
        if not isinstance(self.raw_text, str):
            raise TypeError("TargetSpec rawText must be a string")
        enum_fields = {
            "status": (self.status, TargetSpecStatus),
            "subjectType": (self.subject_type, SubjectType),
            "framing": (self.framing, Framing),
            "source": (self.source, TargetSpecSource),
        }
        for field_name, (value, expected_type) in enum_fields.items():
            if not isinstance(value, expected_type):
                raise TypeError(f"TargetSpec {field_name} must use {expected_type.__name__}")
        if self.status is not TargetSpecStatus.FAILED and not self.raw_text.strip():
            raise ValueError("TargetSpec rawText must not be empty unless status is failed")
        if self.object_label is not None:
            if not isinstance(self.object_label, str):
                raise TypeError("TargetSpec objectLabel must be a string or null")
            if self.schema_version == "0.1":
                raise ValueError("TargetSpec objectLabel requires schemaVersion 0.2")
            if self.subject_type is not SubjectType.OBJECT:
                raise ValueError("TargetSpec objectLabel is only valid for subjectType=object")
            if not OBJECTS365_YOLO26.is_object_label(self.object_label):
                raise ValueError(
                    "TargetSpec objectLabel must be an exact Objects365 canonical object label"
                )
        if self.subject_count is not None and (
            type(self.subject_count) is not int or self.subject_count < 1
        ):
            raise ValueError("TargetSpec subjectCount must be null or an integer of at least 1")
        if isinstance(self.confidence, bool) or not isinstance(self.confidence, (int, float)):
            raise TypeError("TargetSpec confidence must be a number")
        if not math.isfinite(self.confidence) or not 0.0 <= self.confidence <= 1.0:
            raise ValueError("TargetSpec confidence must be in [0, 1]")

    @classmethod
    def from_dict(cls, payload: Mapping[str, object]) -> TargetSpec:
        if not isinstance(payload, Mapping):
            raise TypeError("TargetSpec must be a JSON object")

        allowed_fields = {
            "schemaVersion",
            "sessionId",
            "status",
            "subjectType",
            "objectLabel",
            "subjectCount",
            "framing",
            "rawText",
            "confidence",
            "source",
        }
        unknown_fields = sorted(set(payload) - allowed_fields)
        if unknown_fields:
            raise ValueError(f"Unknown TargetSpec fields: {', '.join(unknown_fields)}")

        for required_field in ("sessionId", "rawText", "source"):
            if required_field not in payload:
                raise ValueError(f"TargetSpec is missing required field: {required_field}")

        # Payloads created before objectLabel existed often omitted the version.
        # Keep interpreting those as v0.1; new v0.2 producers must be explicit.
        schema_version = _string_field(payload, "schemaVersion", default="0.1")
        if schema_version == "0.1" and "objectLabel" in payload:
            raise ValueError("TargetSpec v0.1 must not contain objectLabel")
        session_id = _string_field(payload, "sessionId")
        raw_text = _string_field(payload, "rawText", allow_empty=True)
        object_label = payload.get("objectLabel")
        if object_label is not None and not isinstance(object_label, str):
            raise TypeError("TargetSpec objectLabel must be a string or null")
        subject_count = payload.get("subjectCount")
        if subject_count is not None and (type(subject_count) is not int or subject_count < 1):
            raise ValueError("TargetSpec subjectCount must be null or an integer of at least 1")

        confidence = payload.get("confidence", 0.0)
        if isinstance(confidence, bool) or not isinstance(confidence, (int, float)):
            raise TypeError("TargetSpec confidence must be a number")

        return cls(
            schema_version=schema_version,
            session_id=session_id,
            status=_enum_field(
                TargetSpecStatus,
                payload.get("status", TargetSpecStatus.OK.value),
                "status",
            ),
            subject_type=_enum_field(
                SubjectType,
                payload.get("subjectType", SubjectType.PERSON.value),
                "subjectType",
            ),
            object_label=object_label,
            subject_count=subject_count,
            framing=_enum_field(
                Framing,
                payload.get("framing", Framing.FULL_BODY.value),
                "framing",
            ),
            raw_text=raw_text,
            confidence=float(confidence),
            source=_enum_field(TargetSpecSource, payload["source"], "source"),
        )

    @classmethod
    def from_json(cls, value: str) -> TargetSpec:
        try:
            payload = json.loads(value)
        except json.JSONDecodeError as exc:
            raise ValueError(f"Invalid TargetSpec JSON: {exc.msg}") from exc
        return cls.from_dict(payload)

    @classmethod
    def from_file(cls, path: str | Path) -> TargetSpec:
        target_path = Path(path)
        try:
            contents = target_path.read_text(encoding="utf-8")
        except OSError as exc:
            raise ValueError(f"Could not read TargetSpec file {target_path}: {exc}") from exc
        return cls.from_json(contents)

    def to_dict(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "schemaVersion": self.schema_version,
            "sessionId": self.session_id,
            "status": self.status.value,
            "subjectType": self.subject_type.value,
        }
        if self.schema_version == "0.2":
            payload["objectLabel"] = self.object_label
        payload.update(
            {
                "subjectCount": self.subject_count,
                "framing": self.framing.value,
                "rawText": self.raw_text,
                "confidence": float(self.confidence),
                "source": self.source.value,
            }
        )
        return payload


def _string_field(
    payload: Mapping[str, object],
    name: str,
    *,
    default: str | None = None,
    allow_empty: bool = False,
) -> str:
    value = payload.get(name, default)
    if not isinstance(value, str):
        raise TypeError(f"TargetSpec {name} must be a string")
    if not allow_empty and not value.strip():
        raise ValueError(f"TargetSpec {name} must be a non-empty string")
    return value


def _enum_field(enum_type: type[EnumValue], value: object, name: str) -> EnumValue:
    if not isinstance(value, str):
        raise TypeError(f"TargetSpec {name} must be a string")
    try:
        return enum_type(value)
    except ValueError as exc:
        allowed = ", ".join(item.value for item in enum_type)
        raise ValueError(f"TargetSpec {name} must be one of: {allowed}") from exc

File: frontend/app/src/main/java/com/example/snap_sight/cv/TargetSpec.kt
package com.example.snap_sight.cv

import org.json.JSONObject

/**
 * ① STT/NLU 가 만드는 의도 스펙. `ai/target_spec_schema.md` v0.1/v0.2 의 Kotlin 표현.
 *
 * 후보 선택은 [Objects365TargetSelector] (`ai/on_device_cv/target_selection.py` 포팅)가
 * tracking 뒤에 수행한다. 이 파일은 파싱·검증만 담당하고 선택 규칙을 알지 못한다.
 *
 * 값이 항상 null 일 수 있다는 게 계약이다 — 마이크 권한이 없거나 발화를 건너뛴 세션
 * (`CaptureSessionManager.startSession()` 참고)에서는 의도 자체가 존재하지 않는다.
 */
data class TargetSpec(
    val sessionId: String,
    val rawText: String,
    val source: String,
    val schemaVersion: String = "0.1",
    val status: Status = Status.OK,
    val subjectType: SubjectType = SubjectType.PERSON,
    val objectLabel: String? = null,
    val subjectCount: Int? = null,
    val framing: Framing = Framing.FULL_BODY,
    val confidence: Float = 0f,
) {
    enum class Status(val wire: String) {
        OK("ok"),
        NEEDS_CLARIFICATION("needs_clarification"),
        FAILED("failed");

        companion object {
            fun fromWire(value: String): Status? = entries.firstOrNull { it.wire == value }
        }
    }

    enum class SubjectType(val wire: String) {
        PERSON("person"),
        OBJECT("object"),
        LANDSCAPE("landscape");

        companion object {
            fun fromWire(value: String): SubjectType? = entries.firstOrNull { it.wire == value }
        }
    }

    enum class Framing(val wire: String) {
        CLOSEUP("closeup"),
        FULL_BODY("full_body"),
        WIDE("wide");

        companion object {
            fun fromWire(value: String): Framing? = entries.firstOrNull { it.wire == value }
        }
    }

    init {
        require(schemaVersion in SUPPORTED_SCHEMA_VERSIONS) {
            "Unsupported TargetSpec schemaVersion: $schemaVersion"
        }
        require(sessionId.isNotBlank()) { "TargetSpec sessionId must be a non-empty string" }
        require(status == Status.FAILED || rawText.isNotBlank()) {
            "TargetSpec rawText must not be empty unless status is failed"
        }
        if (objectLabel != null) {
            require(schemaVersion != "0.1") { "TargetSpec objectLabel requires schemaVersion 0.2" }
            require(subjectType == SubjectType.OBJECT) {
                "TargetSpec objectLabel is only valid for subjectType=object"
            }
            require(objectLabel.isNotBlank()) { "TargetSpec objectLabel must not be blank" }
        }
        require(subjectCount == null || subjectCount >= 1) {
            "TargetSpec subjectCount must be null or an integer of at least 1"
        }
        require(confidence.isFinite() && confidence in 0f..1f) {
            "TargetSpec confidence must be in [0, 1]"
        }
    }

    /**
     * 실제 선택 로직이 붙기 전까지 CV 가 이 스펙을 신뢰해도 되는지의 단일 판정 지점.
     * `needs_clarification`/`failed` 는 임의의 객체를 고르면 안 된다는 뜻이다.
     */
    val isActionable: Boolean get() = status == Status.OK

    companion object {
        val SUPPORTED_SCHEMA_VERSIONS = setOf("0.1", "0.2")

        /** 파싱 실패 시 예외를 던진다. 검증 규칙은 `ai/target_spec.py` 와 맞춘다. */
        fun fromJson(json: String): TargetSpec = fromJsonObject(JSONObject(json))

        /**
         * **의도 입력의 null-안전 진입점.**
         *
         * 발화가 없거나(null), 빈 문자열이거나, 스키마를 어긴 payload 가 와도 예외 없이 null 을
         * 돌려준다. CV 루프는 의도 유무와 무관하게 계속 돌아야 하므로 여기서 삼켜야 한다.
         * 무엇 때문에 버렸는지는 [onError] 로 받아 로그에 남긴다.
         */
        fun fromJsonOrNull(json: String?, onError: (Throwable) -> Unit = {}): TargetSpec? {
            if (json.isNullOrBlank()) return null
            return try {
                fromJson(json)
            } catch (t: Throwable) {
                onError(t)
                null
            }
        }

        private fun fromJsonObject(payload: JSONObject): TargetSpec {
            val unknownFields = payload.keys().asSequence().filterNot { it in ALLOWED_FIELDS }.toList()
            require(unknownFields.isEmpty()) {
                "Unknown TargetSpec fields: ${unknownFields.sorted().joinToString(", ")}"
            }
            for (required in REQUIRED_FIELDS) {
                require(payload.has(required)) { "TargetSpec is missing required field: $required" }
            }

            // objectLabel 이 없던 시절 payload 는 버전을 생략한 경우가 많다 → 0.1 로 해석.
            val schemaVersion = payload.optString("schemaVersion", "0.1")
            require(!(schemaVersion == "0.1" && payload.has("objectLabel"))) {
                "TargetSpec v0.1 must not contain objectLabel"
            }

            return TargetSpec(
                schemaVersion = schemaVersion,
                sessionId = payload.getString("sessionId"),
                status = enumField(payload.optString("status", "ok"), "status", Status::fromWire) {
                    Status.entries.map(Status::wire)
                },
                subjectType = enumField(
                    payload.optString("subjectType", "person"),
                    "subjectType",
                    SubjectType::fromWire,
                ) { SubjectType.entries.map(SubjectType::wire) },
                objectLabel = if (payload.isNull("objectLabel")) null else payload.optString("objectLabel"),
                subjectCount = if (payload.isNull("subjectCount")) null else payload.getInt("subjectCount"),
                framing = enumField(
                    payload.optString("framing", "full_body"),
                    "framing",
                    Framing::fromWire,
                ) { Framing.entries.map(Framing::wire) },
                rawText = payload.getString("rawText"),
                confidence = payload.optDouble("confidence", 0.0).toFloat(),
                source = payload.getString("source"),
            )
        }

        private fun <T> enumField(
            value: String,
            name: String,
            parse: (String) -> T?,
            allowed: () -> List<String>,
        ): T = parse(value)
            ?: throw IllegalArgumentException(
                "TargetSpec $name must be one of: ${allowed().joinToString(", ")}"
            )

        private val REQUIRED_FIELDS = listOf("sessionId", "rawText", "source")

        private val ALLOWED_FIELDS = setOf(
            "schemaVersion", "sessionId", "status", "subjectType", "objectLabel",
            "subjectCount", "framing", "rawText", "confidence", "source",
        )
    }
}
